### PR TITLE
Update EndpointSlice from alpha to beta

### PR DESCRIPTION
**Pull request by howardjohn**

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
Just one note, by this, istio can not work on k8s 1.16 if endpointslice enabled.

**Comment by howardjohn** [comment]
I don't think anyone is going to use endpoint slice on 1.16, there are some pretty critical bugs. Is there anyone doing this?

**Comment by hzxuzhonghu** [comment]
maybe some new users/developers, and we can doc is clearly in 1.5 release note.

File: prow/config/endpointslice.yaml [hunk 1]
```diff
@@ -17,5 +17,6 @@ kubeadmConfigPatches:
   controllerManager:
     extraArgs:
       "controllers": "*,endpointslice"
+      "feature-gates": "EndpointSlice=true"
```
Comment (hzxuzhonghu) [inline prow/config/endpointslice.yaml]: Is this for e2e test? I think this can be in a separate pr with any new test

Comment (howardjohn) [reply prow/config/endpointslice.yaml]: it's not used in any e2e test (yet).

**Comment by howardjohn** [comment]
> Just one note, by this, istio can not work on k8s 1.16 if endpointslice enabled

Don't we have the same problem now though, without this users on 1.17 will not work?

**Comment by hzxuzhonghu** [comment]
Without testing, but from the api view, pilot can get endpointslice from kube-apiserver by v1alpha interface.

File: prow/config/endpointslice.yaml (hunk 1 above)
Comment (hzxuzhonghu) [reply prow/config/endpointslice.yaml]: What's the meaning enable it in k8s deploy?

Comment (howardjohn) [reply prow/config/endpointslice.yaml]: yes this is the config to turn on endpoint slice. The change is needed because it 1.17 they made some changes. It's not used for tests but I am going to add a test soon

**Comment by howardjohn** [comment]
So of you query for alpha you get alpha+beta but if you query for beta you just get alpha?

I think at one point I had code query for alpha but was inserting beta objects and it returned no results

**Comment by hzxuzhonghu** [comment]
>So of you query for alpha you get alpha+beta but if you query for beta you just get alpha?

No, if one deployed k8s 1.16 with endpointslice enabled, and a client requesting v1beta, there will be an error indicating no such group/version resource? However, if a client request v1alpha1 in 1.17, then it can get both v1alpha and v1beta resources. Kube-apiserver does the conversion from an internal version to requested version.

**Comment by howardjohn** [comment]
> No, if one deployed k8s 1.16 with endpointslice enabled, and a client requesting v1beta, there will be an error indicating no such group/version resource? However, if a client request v1alpha1 in 1.17, then it can get both v1alpha and v1beta resources. Kube-apiserver does the conversion from an internal version to requested version.

Ack, I verified this in a real cluster just to double check. 1.16 will be broken if we make this change

@nmittler did you need features that are in the beta api?

**Comment by nmittler** [comment]
@howardjohn I don't think so ... alpha is fine for now. Do we know when k8s will stop shipping with the alpha version of the CRD?

**Comment by howardjohn** [comment]
Not sure. I don't think there is any need to support kube-controller
endpointslice in 1.16 since its very alpha, but people may want to manually
make endpointslices in 1.16 so might as well keep alpha around unless we
have a solid reason not to?

**Comment by hzxuzhonghu** [comment]
I would agree, we keep using alpha client, it is compatible even k8s bumped to beta. Unless one day, k8s deprecated alpha version. That is to say when k8s 1.16 is not maintained(maybe when 1.19 released), we can upgrade to beta.  Then we should also declare istio stops supporting k8s 1.16.

**Comment by howardjohn** [comment]
Close per ^

**Comment by brianwolfe** [comment]
This ends up requiring that you enable the alpha API as well, which is not enabled by default. I tested this with a Kind cluster running stock 1.18:

```
kind create cluster --config ./istio-testing.yaml -v9 --retain --image kindest/node:v1.18.2 --wait=60s
```

cat <<EOF > cluster.yaml
---
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
  - role: control-plane
  - role: worker
kubeadmConfigPatches:
  - |
    apiVersion: kubeadm.k8s.io/v1beta2
    kind: ClusterConfiguration
    metadata:
      name: config
    apiServer:
      extraArgs:
        "service-account-issuer": "kubernetes.default.svc"
        "service-account-signing-key-file": "/etc/kubernetes/pki/sa.key"
containerdConfigPatches:
  - |-
    [plugins."io.containerd.grpc.v1.cri".registry.mirrors."localhost:5000"]
      endpoint = ["http://kind-registry:5000"]
EOF

kind create cluster --config ./config.yaml -v9 --retain --image kindest/node:v1.18.2 --wait=60s
```

Then create an Istiod deployment with EndpointSlice enabled, I see
```
2021-01-05T01:01:59.005682Z	error	k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125: Failed to list *v1alpha1.EndpointSlice: the server could not find the requested resource
```

**Comment by howardjohn** [comment]
We use v1beta1 in 1.8

**Comment by brianwolfe** [comment]
Thanks, you are right, I was testing on 1.7.5 and got a bit confused.
